src/main/telemetry/sim.c

```diff
@@ -0,0 +1,409 @@
+        // ERROR
+        atCommandStatus = SIM_AT_ERROR;
+        if (!simWaitAfterResponse) {
+            sim_t_stateChange = millis() + SIM_AT_COMMAND_DELAY_MIN_MS;
+        }
+        return;
+    } else if (responseCode == SIM_RESPONSE_CODE_RING) {
+        // RING
+    } else if (responseCode == SIM_RESPONSE_CODE_CSQ) {
+        // +CSQ: 26,0
+        simRssi = fastA2I((char*)&simResponse[6]);
+    } else if (responseCode == SIM_RESPONSE_CODE_CLIP) {
+        // we always get this after a RING when a call is incoming
+        // +CLIP: "+3581234567"
+        readOriginatingNumber(&simResponse[8]);
+        if (checkGroundStationNumber(&simResponse[8])) {
+            requestSendSMS(SIM_TX_FLAG_RESPONSE);
+        }
+    } else if (responseCode == SIM_RESPONSE_CODE_CREG) {
+        // +CREG: 0,1
+        if (simResponse[9] == '1' || simResponse[9] == '5') {
+            simModuleState = SIM_MODULE_REGISTERED;
+        } else {
+            simModuleState = SIM_MODULE_NOT_REGISTERED;
+        }        
+    } else if (responseCode == SIM_RESPONSE_CODE_CMT) {
+        // +CMT: <oa>,[<alpha>],<scts>[,<tooa>,<fo>,<pid>,<dcs>,<sca>,<tosca>,<length>]<CR><LF><data>
+        // +CMT: "+3581234567","","19/02/12,14:57:24+08"
+        readOriginatingNumber(&simResponse[7]);
+        if (checkGroundStationNumber(&simResponse[7])) {
+            readState = SIM_READSTATE_SMS; // next simResponse line will be SMS content
+        } else {
+            readState = SIM_READSTATE_SKIP; // skip SMS content
+        }
+    }
+}
+
+void detectAccEvents()
+{
+    timeMs_t timeSinceMsg = millis() - t_lastMessageSent;
+
+    if (lastMessageTriggeredBy == SIM_TX_FLAG_ACC && timeSinceMsg < 2000)
+        return;
+
+    uint32_t accSq = sq(imuMeasuredAccelBF.x) + sq(imuMeasuredAccelBF.y) + sq(imuMeasuredAccelBF.z);
+//    float acceleration = sqrtf(vectorNormSquared(acc.accADCf));
+
+    if (telemetryConfig()->accEventThresholdHigh > 0 && accSq > sq(telemetryConfig()->accEventThresholdHigh))
+        accEvent = ACC_EVENT_HIGH;
+    else if (accSq < sq(telemetryConfig()->accEventThresholdLow))
+        accEvent = ACC_EVENT_LOW;
+    else if (telemetryConfig()->accEventThresholdNegX > 0 && imuMeasuredAccelBF.x < -telemetryConfig()->accEventThresholdNegX)
+        accEvent = ACC_EVENT_NEG_X;
+    else
+        return;
+
+    if (lastMessageTriggeredBy != SIM_TX_FLAG_ACC || timeSinceMsg > 1000 * telemetryConfig()->simTransmitInterval) {
+        requestSendSMS(SIM_TX_FLAG_ACC);
+    }
+}
+
+void transmit()
+{
+    timeMs_t timeSinceMsg = millis() - t_lastMessageSent;
+
+    if (timeSinceMsg < 1000 * MAX(SIM_MIN_TRANSMIT_INTERVAL, telemetryConfig()->simTransmitInterval))
+        return;
+
+    uint8_t triggers = SIM_TX_FLAG;
+
+    if (FLIGHT_MODE(FAILSAFE_MODE))
+        triggers |= SIM_TX_FLAG_FAILSAFE;
+    if (gpsSol.fixType != GPS_NO_FIX && gpsSol.numSat < gpsConfig()->gpsMinSats)
+        triggers |= SIM_TX_FLAG_GPS;
```

Comment:
This is not particularly useful, there are also .
I suggest to use something like `bool navigationPositionEstimateIsHealthy()`
which returns:  `(posControl.flags.estPosStatus >= EST_USABLE) && STATE(GPS_FIX_HOME);`
And send sms when fix was and now it is gone:
`STATE(GPS_FIX_HOME) && (posControl.flags.estPosStatus < EST_USABLE)`
`estPosStatus` is always checking in `navigation.c`

Reply:
To send a message when fix is lost, we must store the last good GPS position. Is this available somewhere or should we add variables for it?

Reply:
Hm.. this condition:
`posControl.flags.estPosStatus >= EST_USABLE`
 basically means:
`posEstimator.est.eph < positionEstimationConfig()->max_eph_epv`
if I'm reading the code correctly.
So we would start sending messages when position uncertainty grows to more than 10 meters?

Reply:
Yes that's it. The last trusted location would have to be saved in a variable.